README.txt
Introduction
------------

*mfabrik.zoho* package provides Python classes for making easy Zoho API calls. 
They allow you to integrate `Zoho web application data <http://www.zoho.com>`_
into your Python software.  

Features
--------

* Creating Zoho API session a.k.a ticket

* Extendable API using a base class architecture

* Logging all incoming and outgoing API traffic with Zoho

* Support Python 2.4, 2.5 and 2.6 (2.4 needs additional lxml and simplejson libraries installed)

* Compatible with non-ASCII or Unicode letters

* Unit tests to guarantee the high quality of the code

*mfabrik.zoho* is intended to use for machine-to-machine communication and
will work with any web framework, like Plone, Django, Google App Engine.

To communicate with Zoho you need username, password and API KEY.
For further information, see *Setup > Admin > Developer key* in 
your Zoho application.

API support
-----------

Currently out of box support includes:

* CRM apis: insert_records, get_records, delete_lead

You can easily wrap Zoho API calls you need using this library.
Please contribute patches to the package.

Installation
------------

To install mfabrik.zoho to your system-wide Python 
do as below. 

Example (UNIX)::

        sudo easy_install mfabrik.zoho
        
Example (UNIX, Python 2.4)::

        sudo easy_install mfabrik.zoho lxml simplejson 
        
Usage
-----

To learn how to use this library, it is best to study its unit test source code (tests.py).

Example usage::

        # Import CRM connector class
        from mfabrik.zoho.crm import CRM
        from mfabrik.zoho.core import ZohoException

        # Initialize Zoho CRM API connection
        # You need valid Zoho credentials and API key for this.
        # You get necessary data from Zoho > Setup > Admin > Developer key
        crm = CRM(username="myusername", password="foobar", apikey="12312312312312312323")

        # Open connection can be used to make as many as possible API calls
        # This will raise ZohoException if credentials are incorrect.
        # Also IOError or URLError will be raised if you the connection to Zoho servers
        # does not work.
        crm.open()

        # Lead is just a bunch of dictionaried data
        # For possible lead parameters see crm.py.
        
        # Zoho default compulsory fields: Last Name, Company
        
        lead = {
            u"First Name" : u"Mikko",
            u"Last Name" : u"Ohtamaa",
            u"Company" : u"mFabrik Research Oy"   
        }

        # Insert a new lead to Zoho CRM lead database.
        # We support multiple leads per call, so we need to listify our one lead first.
        responses = crm.insert_records([lead]) # This will raise ZohoException if lead data is invalid
        
        # list of responses. one response is {'Modified Time': '2010-10-07 13:24:49', 'Created By': 'Developer', 'Modified By': 'Developer', 'Id': '177376000000253053', 'Created Time': '2010-10-07 13:24:49'}
        # At least one response is guaranteed, otherwise an exception is raised
        
        lead_id = responses[0]["Id"]
        

Special field *Lead Owner* must be the registered email fo Zoho CRM user.        
        
        
.. note::
        
        Some calls (e.g. delete) seem to have delay and the changes might not be instantly
        reflected in the next API call (getRecords).
        
        
Logging
=======

Python `logging` module logger *Zoho API* is used to output API traffic
on DEBUG log level.

Source code
-----------

* http://github.com/miohtama/mfabrik.zoho

Commercial development
-----------------------

This package is licensed under open source GPL license.
If you wish to use this code in a commercial product, 
relicense it or you are 
looking for high quality Zoho/Python support, please contact
`mFabrik Research <mailto:research@mfabrik.com>`_.
Our top class Python developers are ready to help you with your software development.

Further reading
---------------

* Zoho CRM API: http://zohocrmapi.wiki.zoho.com/API-Methods.html

* API update notes: http://forums.zoho.com/topic/zoho-crm-api-update-important

Author
------

`mFabrik Research Oy <mailto:info@mfabrik.com>`_ - Python and Plone professionals for hire.

* `mFabrik web site <http://mfabrik.com>`_ 

* `mFabrik mobile site <http://mfabrik.mobi>`_ 

* `Blog <http://blog.mfabrik.com>`_

       

docs/HISTORY.txt
Changelog
=========

1.0.2 - 1.1
------------------

* Changed protocol to https, as per Zoho API changes [miohtama]

1.0.1 - 1.0.2
-------------

* Return CRM lead insert id results [tco]

1.0 - 1.0.1
-----------

* Fixed MANIFEST.in file - setuptools was ignoring docs/ by default [miohtama]


1.0
-------------------

- Initial release


mfabrik/zoho/tests.py
# -*- coding: utf-8 -*-
"""

    Zoho API bridge unit tests.
    
    These unit-test assume you have set-up your user account specific environment varibles:
    
    * ZOHO_USERNAME
    
    * ZOHO_PASSWORD
    
    * ZOHO_APIKEY
    
    Since you are probably using proudction Zoho database to testing, 
    all tests include clean-up functions which should clear all test data
    if tests are succesfully completed.

"""

__copyright__ = "2010 mFabrik Research Oy"
__author__ = "Mikko Ohtamaa <mikko@mfabrik.com>"
__license__ = "GPL"
__docformat__ = "Epytext"

import os, sys
import unittest

#import mfabrik
#from mfabrik import zoho
from crm import CRM
from core import ZohoException
     
def enable_debug_log():
    """ Enable context.plone_log() output from Python scripts """
    import sys, logging
    from Products.CMFPlone.log import logger
    logger.root.setLevel(logging.DEBUG)
    logger.root.addHandler(logging.StreamHandler(sys.stdout))

enable_debug_log()

class TestCRM(unittest.TestCase):
    """ Zoho CRM specific tests. """

    def setUp(self):
        
        
        self.username = os.environ.get("ZOHO_USERNAME", None)
        self.password = os.environ.get("ZOHO_PASSWORD", None)
        self.apikey = os.environ.get("ZOHO_APIKEY", None)
        
        if self.username and self.password and self.apikey:
            # All ok
            pass
        else:
            raise RuntimeError("Please set-up unit-test environment variables: ZOHO_USERNAME, ZOHO_PASSWORD and ZOHO_APIKEY to run tests")

        # Initialize Zoho CRM api connection
        # You get necessary data from Zoho > Setup > Admin > Developer key
        self.crm = CRM(username=self.username, password=self.password, apikey=self.apikey)

        # Open connection can be used to make as many as possible API calls
        self.crm.open()
        
        self.clear_test_leads()
        
    def tearDown(self):
        self.clear_test_leads()
        

    def add_dummy_lead(self):

        # Lead is just a bunch of dictionaried data
        # For possible lead parameters see crm.py.
        # Include non-ASCII characters.
        
        lead = {
            u"First Name" : u"TEST",
            u"Last Name" : u"UNIT TEST ÅÄÖ",
            u"Company" : u"mFabrik Research Oy"   
        }

        # Insert a new lead to Zoho CRM lead database.
        # We support multiple leads per call, so we need to listify our one lead first.
        self.crm.insert_records([lead])
        
    def clear_test_leads(self):
        """ Remove all test leads from the database """
        
        records = self.crm.get_records()
        
        for record in records:
            if record["First Name"] == "TEST":
                self.crm.delete_record(record["LEADID"])
        
        
    def test_insert_lead(self):
        """ Sample insert lead scenario. """

        self.add_dummy_lead()

        
    def xxx_test_insert_duplicate(self):
        """ What happens when a duplicate lead is inserted """

        self.add_dummy_lead()
        self.add_dummy_lead()

        # TODO: Need to figure out how to force Zoho correctly detect duplicates
        
    def test_insert_missing_field(self):
        """ Company and Last name are required lead fields """

        
        lead = {
            "First Name" : "Who?"
        }

        try:
            self.crm.insert_records([lead])
            raise AssertionError("Should not be reached")
        except ZohoException:
            pass
        
    def test_get_leads(self):

        # Make sure we have add least one lead there
        self.add_dummy_lead()
        
        leads = self.crm.get_records()
        print "Got leads"
        print leads
        
    def test_delete_leads(self):

        # Make sure we have add least one lead there
        self.add_dummy_lead()
        
        records = self.crm.get_records()
        
        # Count number of test items
        counter = 0
        for r in records:
            if r["First Name"] == "TEST":
                counter += 1
                
        self.assertEqual(counter, 1)
        

def test_suite():
    from unittest import TestSuite, makeSuite
    suite = TestSuite()
    suite.addTest(makeSuite(TestCRM))
    return suite

if __name__ == '__main__':
    unittest.main()

mfabrik/zoho/crm.py
"""

    Zoho CRM API bridge.

"""

__copyright__ = "2010 mFabrik Research Oy"
__author__ = "Mikko Ohtamaa <mikko@mfabrik.com>"
__license__ = "GPL"
__docformat__ = "Epytext"


try:
    from xml import etree
    from xml.etree.ElementTree import Element, tostring, fromstring
except ImportError:
    try:
        from lxml import etree
        from lxml.etree import Element, tostring, fromstring
    except ImportError:
        raise RuntimeError("XML library not available:  no etree, no lxml")
   
from core import Connection, ZohoException, decode_json

class CRM(Connection):
    """ CRM specific Zoho APIs mapped to Python """
    
    def get_service_name(self):
        """ Called by base class """
        return "ZohoCRM"
    
    def check_successful_xml(self, response):
        """ Make sure that we get "succefully" response.
        
        Throw exception of the response looks like something not liked.
        
        @raise: ZohoException if any error
        
        @return: Always True
        """

        # Example response
        # <response uri="/crm/private/xml/Leads/insertRecords"><result><message>Record(s) added successfully</message><recorddetail><FL val="Id">177376000000142007</FL><FL val="Created Time">2010-06-27 21:37:20</FL><FL val="Modified Time">2010-06-27 21:37:20</FL><FL val="Created By">Ohtamaa</FL><FL val="Modified By">Ohtamaa</FL></recorddetail></result></response>

        root = fromstring(response)
            
        # Check error response
        # <response uri="/crm/private/xml/Leads/insertRecords"><error><code>4401</code><message>Unable to populate data, please check if mandatory value is entered correctly.</message></error></response>
        for error in root.findall("error"):
            print "Got error"
            for message in error.findall("message"):
                raise ZohoException(message.text)
        
        return True
    
    def insert_records(self, leads, extra_post_parameters={}):
        """ Insert new leads to Zoho CRM database.
        
        The contents of the lead parameters can be defined in Zoho CRM itself.
        
        http://zohocrmapi.wiki.zoho.com/insertRecords-Method.html
        
        @param leads: List of dictionaries. Dictionary content is directly mapped to 
            <FL> XML parameters as described in Zoho CRM API.
        
        @param extra_post_parameters: Parameters appended to the HTTP POST call. 
            Described in Zoho CRM API.
        
        @return: List of record ids which were created by insert recoreds
        """        

        self.ensure_opened()
        
        root = Element("Leads")

        # Row counter
        no = 1

        for lead in leads:
            row = Element("row", no=str(no))
            root.append(row)

            assert type(lead) == dict, "Leads must be dictionaries inside a list, got:" + str(type(lead))
        
            for key, value in lead.items():
                # <FL val="Lead Source">Web Download</FL>
                # <FL val="First Name">contacto 1</FL>
                fl = Element("fl", val=key)
                fl.text = value
                row.append(fl)
                
            no += 1

        post = {
            'newFormat':    1,
            'duplicateCheck':   2
        }

        post.update(extra_post_parameters)
        
        response = self.do_xml_call("https://crm.zoho.com/crm/private/xml/Leads/insertRecords", post, root)

        self.check_successful_xml(response)
                
        return self.get_inserted_records(response)
        
    
    def get_inserted_records(self, response):
        """
        @return: List of record ids which were created by insert recoreds
        """
        root = fromstring(response)
        
        records = []
        for result in root.findall("result"):
            for record in result.findall("recorddetail"):
                record_detail = {}
                for fl in record.findall("FL"):
                    record_detail[fl.get("val")] = fl.text
                records.append(record_detail)
        return records
        
    
    def get_records(self, selectColumns='leads(First Name,Last Name,Company)', parameters={}):
        """ 
        
        http://zohocrmapi.wiki.zoho.com/getRecords-Method.html
        
        @param selectColumns: String. What columns to query. For example query format,
            see API doc. Default is leads(First Name,Last Name,Company).
        
        @param parameters: Dictionary of filtering parameters which are part of HTTP POST to Zoho.
            For example parameters see Zoho CRM API docs.
        
        @return: Python list of dictionarizied leads. Each dictionary contains lead key-value pairs. LEADID column is always included.

        """
        
        self.ensure_opened()
        

        post_params = {
            "selectColumns" : selectColumns,
            "newFormat" : 2
        }
        
        post_params.update(parameters)

        response = self.do_call("https://crm.zoho.com/crm/private/json/Leads/getRecords", post_params)
        
        # raw data looks like {'response': {'result': {'Leads': {'row': [{'FL': [{'content': '177376000000142085', 'val': 'LEADID'}, ...
        data =  decode_json(response)
        
        # Sanify output data to more Python-like format
        output = []
        for row in data["response"]["result"]["Leads"]["row"]:
            item = {}
            for cell in row["FL"]:
                item[cell["val"]] = cell["content"]
            
            output.append(item)
            
        return output
        
                
    def delete_record(self, id, parameters={}):
        """ Delete one record from Zoho CRM.
                        
        @param id: Record id
        
        @param parameters: Extra HTTP post parameters        
        """
        
        self.ensure_opened()
    
        post_params = {}
        post_params["id"] = id
        post_params.update(parameters)
        
        response = self.do_call("https://crm.zoho.com/crm/private/xml/Leads/deleteRecords", post_params)
        
        self.check_successful_xml(response)
        

mfabrik/zoho/core.py
"""

    Zoho API core functions.

"""

__copyright__ = "2010 mFabrik Research Oy"
__author__ = "Mikko Ohtamaa <mikko@mfabrik.com>"
__license__ = "GPL"
__docformat__ = "Epytext"

import urllib, urllib2

import logging

try:
    from xml import etree
    from xml.etree.ElementTree import Element, tostring, fromstring
except ImportError:
     try:
         from lxml import etree
         from lxml.etree import  Element, tostring, fromstring
     except ImportError:
         print "XML library not available:  no etree, no lxml"
         raise
   

try:
    import json
except ImportError:
    try:
        import simplejson
    except ImportError:
        # Python 2.4, no simplejson installed
        raise RuntimeError("You need json or simplejson library with your Python")


logger = logging.getLogger("Zoho API")


class ZohoException(Exception):
    """ Bad stuff happens. 
    
    If it's level 15 or higher bug, you usually die.
    If it's lower level then you just lose all your data.
    
    Play some Munchkin.
    """


class Connection(object):
    """ Zoho API connector.
        
    Absract base class for all different Zoho API connections.
    Subclass this and override necessary methods to support different Zoho API groups.
    """

    
    def __init__(self, username, password, apikey, extra_auth_params = {}, auth_url="https://accounts.zoho.com/login"):        
        """
        @param username: manifisto@mfabrik.com 
        
        @param password: xxxxxxx
        
        @param apikey: Given by Zoho, string like 123123123-rVI20JVBveUOHIeRYWV5b5kQaMGWeIdlI$
                
        @param extra_auth_params: Dictionary of optional HTTP POST parameters passed to the login call
        
        
        @param auth_url: Which URL we use for authentication
        """        
        self.username = username
        self.password = password
        self.apikey = apikey
        # 
        self.auth_url = None
        
        # Ticket is none until the conneciton is opened
        self.ticket = None
        
    def get_service_name(self):
        """ Return API name which we are using. """
        raise NotImplementedError("Subclass must implement")
        
    def open(self):
        """ Open a new Zoho API session """
        self.ticket = self._create_ticket()
    
    def _create_ticket(self):
        """ 
        Ticket idenfities Zoho session.
        
        It is a bit like cookie authentication.
        
        @return: Ticket code
        """
        #servicename=ZohoCRM&FROM_AGENT=true&LOGIN_ID=Zoho Username or Email Address&PASSWORD=Password
        params = {
            'servicename': self.get_service_name(),
            'FROM_AGENT': 'true',
            'LOGIN_ID': self.username,
            'PASSWORD': self.password
        }
        
        requestUrl = "https://accounts.zoho.com/login?%s" % (urllib.urlencode(params))
        request = urllib2.Request(requestUrl, None)
        body = urllib2.urlopen(request).read()
        
        data = self._parse_ticket_response(body)
        
        if data["WARNING"] != "null":
            # Zoho has set an error field
            raise ZohoException("Could not auth:" + data["WARNING"])
        
        if data["RESULT"] != "TRUE":
            raise ZohoException("Ticket result was not valid")

        return data["TICKET"]
    
    def _parse_ticket_response(self, data):
        """ Dictionarize ticket opening response
       
        Example response::
        
            # #Sun Jun 27 20:10:30 PDT 2010 GETUSERNAME=null WARNING=null PASS_EXPIRY=-1 TICKET=3bc26b16d97473a1245dbf93a5dcd153 RESULT=TRUE 
        """
        
        output = {}
        
        lines = data.split("\n")
        for line in lines:
            
            if line.startswith("#"):
                # Comment
                continue
            
            if line.strip() == "":
                # Empty line
                continue
            
            if not "=" in line:
                raise ZohoException("Bad ticket data:" + data)
        
            key, value = line.split("=")
            output[key] = value
            
        return output

    def ensure_opened(self):
        """ Make sure that the Zoho Connection is correctly opened """
        
        if self.ticket is None:
            raise ZohoException("Need to initialize Zoho ticket first")
        
    def do_xml_call(self, url, parameters, root):
        """  Do Zoho API call with outgoing XML payload.
        
        Ticket and apikey parameters will be added automatically.
        
        @param url: URL to be called
        
        @param parameters: Optional POST parameters. 
        
        @param root: ElementTree DOM root node to be serialized.
        """
        
        parameters = parameters.copy()
        parameters["xmlData"] = tostring(root)
        return self.do_call(url, parameters)

    def do_call(self, url, parameters):
        """ Do Zoho API call.
        
        @param url: URL to be called
        
        @param parameters: Optional POST parameters. 
        """
        # Do not mutate orginal dict
        parameters = parameters.copy()
        parameters["ticket"] = self.ticket
        parameters["apikey"] = self.apikey
        
        stringify(parameters)
        
        if logger.getEffectiveLevel() == logging.DEBUG:                                          
            # Output Zoho API call payload
            logger.debug("Doing ZOHO API call:" + url)
            for key, value in parameters.items():
                logger.debug(key + ": " + value)
                
        request = urllib2.Request(url, urllib.urlencode(parameters))
        response = urllib2.urlopen(request).read()

        if logger.getEffectiveLevel() == logging.DEBUG:                                          
            # Output Zoho API call payload
            logger.debug("ZOHO API response:" + url)
            logger.debug(response)
        
        return response

def stringify(params):
    """ Make sure all params are urllib compatible strings """
    for key, value in params.items():
        
        if type(value) == str:
            params[key] == value.decode("utf-8")
        elif type(value) == unicode:
            pass
        else:
            # call __unicode__ of object
            params[key] = unicode(value)
            

def decode_json(json_data):
    """ Helper function to handle Zoho specific JSON decode.

    @return: Python dictionary/list of incoming JSON data
    
    @raise: ZohoException if JSON'ified error message is given by Zoho
    """
    
    # {"response": {"uri":"/crm/private/json/Leads/getRecords","error": {"code":4500,"message":"Problem occured while processing the request"}}}
    data = simplejson.loads(json_data)
    
    response = data.get("response", None)
    if response:
        error = response.get("error", None)
        if error:
            raise ZohoException("Error while calling JSON Zoho api:" + str(error))
    
    return data